File: medium/207. Course Schedule.py
# 207. Course Schedule
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        '''
        1/2 Topological sorting
        '''
        preCount = Counter()
        successors = defaultdict(list)
        for a, b in prerequisites:
            preCount[a] += 1
            successors[b].append(a)
        
        taken = 0
        free = set(range(numCourses)) - preCount.keys()
        while free:
            b = free.pop()
            taken += 1
            for a in successors[b]:
                preCount[a] -= 1
                if preCount[a] == 0:
                    free.add(a)
        
        return taken == numCourses
        '''
        2/2 DFS.
        Build the graph, and ensure no cycles.
        Use DFS with memoization to find cycles.
        Total time is O(E + V) where E is number of edges, V is number of vertices.
        '''
        prereqs = defaultdict(list)
        
        for b, a in prerequisites:
            prereqs[b].append(a)

        path = [False] * numCourses
        
        @lru_cache(None)
        def has_cycle(current):
            if path[current]:
                return True

            path[current] = True

            if any(has_cycle(children) for children in prereqs[current]):
                return True

            path[current] = False
            return False
        
        return not any(has_cycle(current) for current in range(numCourses))

File: medium/207. Course Schedule.rb
# 207. Course Schedule
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
    pre_count = Hash.new(0)
    succs = Hash.new { |h, k| h[k] = [] }
    prerequisites.each do |b, a|
        pre_count[b] += 1
        succs[a].push(b)
    end
    free = Set.new(0...num_courses) - pre_count.keys
    taken = 0
    until free.empty?
        a = free.pop
        taken += 1
        succs[a].each do |b|
            pre_count[b] -= 1
            free.add(b) if pre_count[b] == 0
        end
    end

    taken == num_courses
end

class Set
  def pop
    element = first
    delete(first)
    element
  end
end